src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,487 @@
+                }
+            }
+
+            return activePods.Count;
+        }
+
+        private async Task RunAndRenew(CancellationTokenSource cts)
+        {
+            // in the background we will continually update partition ownership until
+            // we receive a cancellation token to shutdown the processor
+            _logger.LogTrace($"Running the processor with partitions: {string.Join(", ", _ownedPartitions.ToArray())}");
+
+            try
+            {
+                await RenewOwnership(cts.Token);
+                await CheckProcessorCount(cts);
+                await StartAssignedPartitionProcessor(cts.Token, _ownedPartitions);
+            }
+            catch (TaskCanceledException)
+            {
+                _logger.LogTrace("A task was cancelled");
+            }
+        }
+
+        protected override async Task ExecuteAsync(CancellationToken ct)
+        {
+            CreateContainerClient();
+            await RegisterProcessor(ct);
+            await ContinuallyRegisterProcessor(ct);
+            await CheckIfPartitionsAreUnclaimed(ct);
+
+            while (!ct.IsCancellationRequested)
+            {
+                using var innerCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
+                while (!innerCts.IsCancellationRequested)
+                {
+                    await Task.Delay(5000, innerCts.Token);
+                    _eventHubPartitions = await GetPartitionsIdsForEventHub(innerCts.Token);
+                    await AcquirePartitions(innerCts.Token);
+                    await RunAndRenew(innerCts);
+                }
+            }
+        }
+
+        private Task ContinuallyRegisterProcessor(CancellationToken ct)
+        {
+            Task.Run(() => RegisterProcessorLoop(ct), ct);
+            return Task.CompletedTask;
+        }
+
+        private Task CheckProcessorCount(CancellationTokenSource cts)
+        {
+            Task.Run(async () =>
+            {
+                while (!cts.IsCancellationRequested)
+                {
+                    var activePods = ListBlobs(_containerClient, "activepods");
+                    var activeProcessors = await GetActiveProcessors(activePods, _containerClient, cts.Token);
+
+                    _logger.LogTrace($"CheckProcessorCount - active processors {activeProcessors}");
+
+                    if (activeProcessors != _totalProcessorsRunning)
+                    {
+                        _logger.LogTrace($"Detected that number of processors equals {activeProcessors} but this processor is running with {_totalProcessorsRunning} known processors");
+                        _logger.LogTrace("Restarting and recomputing");
+                        cts.Cancel();
+                    }
+
+                    await Task.Delay(30000, cts.Token);
+                }
+            });
+
+            return Task.CompletedTask;
+        }
+
+        private Task CheckIfPartitionsAreUnclaimed(CancellationToken ct)
```

Comment:
Every worker would be doing this correct?  Instead can we emit the delay when we renew?  The gap is if no one owns the partition, nothing will update the metric.  Might be worth investigating if our alerting can detect no events in a time period.

As mentioned in another comment, freshness (the time last renewed) would be better if it weren't for the metric limitations.

Alternatively, could introduce a singleton blob location for a leader and only the leader is responsible for recording the freshness while others just loop and try to claim ownership.